File: src/main/java/cz/vhromada/catalog/facade/impl/SeasonFacadeImpl.java
package cz.vhromada.catalog.facade.impl;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import cz.vhromada.catalog.entity.Season;
import cz.vhromada.catalog.entity.Show;
import cz.vhromada.catalog.facade.SeasonFacade;
import cz.vhromada.catalog.utils.CatalogUtils;
import cz.vhromada.common.facade.AbstractMovableChildFacade;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.utils.CollectionUtils;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.converter.Converter;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * A class represents implementation of facade for seasons.
 *
 * @author Vladimir Hromada
 */
@Component("seasonFacade")
public class SeasonFacadeImpl extends AbstractMovableChildFacade<Season, cz.vhromada.catalog.domain.Season, Show, cz.vhromada.catalog.domain.Show>
    implements SeasonFacade {

    /**
     * Creates a new instance of SeasonFacadeImpl.
     *
     * @param showService     service for shows
     * @param converter       converter
     * @param showValidator   validator for show
     * @param seasonValidator validator for season
     * @throws IllegalArgumentException if service for shows is null
     *                                  or converter is null
     *                                  or validator for show is null
     *                                  or validator for season is null
     */
    @Autowired
    public SeasonFacadeImpl(final MovableService<cz.vhromada.catalog.domain.Show> showService, final Converter converter,
        final MovableValidator<Show> showValidator, final MovableValidator<Season> seasonValidator) {
        super(showService, converter, showValidator, seasonValidator);
    }

    @Override
    protected cz.vhromada.catalog.domain.Season getDomainData(final Integer id) {
        final List<cz.vhromada.catalog.domain.Show> shows = getMovableService().getAll();
        for (final cz.vhromada.catalog.domain.Show show : shows) {
            for (final cz.vhromada.catalog.domain.Season season : show.getSeasons()) {
                if (id.equals(season.getId())) {
                    return season;
                }
            }
        }

        return null;
    }

    @Override
    protected List<cz.vhromada.catalog.domain.Season> getDomainList(final Show parent) {
        return getMovableService().get(parent.getId()).getSeasons();
    }

    @Override
    protected cz.vhromada.catalog.domain.Show getForAdd(final Show parent, final cz.vhromada.catalog.domain.Season data) {
        final cz.vhromada.catalog.domain.Show show = getMovableService().get(parent.getId());
        if (show.getSeasons() == null) {
            show.setSeasons(new ArrayList<>());
        }
        show.getSeasons().add(data);

        return show;
    }

    @Override
    protected cz.vhromada.catalog.domain.Show getForUpdate(final Season data) {
        final cz.vhromada.catalog.domain.Show show = getShow(data);
        final cz.vhromada.catalog.domain.Season season = getDataForUpdate(data);
        season.setEpisodes(getSeason(data.getId(), show).getEpisodes());

        updateSeason(show, season);

        return show;
    }

    @Override
    protected cz.vhromada.catalog.domain.Show getForRemove(final Season data) {
        final cz.vhromada.catalog.domain.Show show = getShow(data);

        final List<cz.vhromada.catalog.domain.Season> seasons = show.getSeasons().stream()
            .filter(seasonValue -> !seasonValue.getId().equals(data.getId()))
            .collect(Collectors.toList());
        show.setSeasons(seasons);

        return show;
    }

    @Override
    protected cz.vhromada.catalog.domain.Show getForDuplicate(final Season data) {
        final cz.vhromada.catalog.domain.Show show = getShow(data);
        final cz.vhromada.catalog.domain.Season seasonDomain = getSeason(data.getId(), show);

        final cz.vhromada.catalog.domain.Season newSeason = CatalogUtils.duplicateSeason(seasonDomain);
        show.getSeasons().add(newSeason);

        return show;
    }

    @Override
    protected cz.vhromada.catalog.domain.Show getForMove(final Season data, final boolean up) {
        final cz.vhromada.catalog.domain.Show show = getShow(data);
        final cz.vhromada.catalog.domain.Season season = getSeason(data.getId(), show);
        final List<cz.vhromada.catalog.domain.Season> seasons = CollectionUtils.getSortedData(show.getSeasons());

        final int index = seasons.indexOf(season);
        final cz.vhromada.catalog.domain.Season other = seasons.get(up ? index - 1 : index + 1);
        final int position = season.getPosition();
        season.setPosition(other.getPosition());
        other.setPosition(position);

        updateSeason(show, season);
        updateSeason(show, other);

        return show;
    }

    @Override
    protected Class<Season> getEntityClass() {
        return Season.class;
    }

    @Override
    protected Class<cz.vhromada.catalog.domain.Season> getDomainClass() {
        return cz.vhromada.catalog.domain.Season.class;
    }

    /**
     * Returns show for season.
     *
     * @param season season
     * @return show for season
     */
    private cz.vhromada.catalog.domain.Show getShow(final Season season) {
        for (final cz.vhromada.catalog.domain.Show show : getMovableService().getAll()) {
            for (final cz.vhromada.catalog.domain.Season seasonDomain : show.getSeasons()) {
                if (season.getId().equals(seasonDomain.getId())) {
                    return show;
                }
            }
        }

        throw new IllegalStateException("Unknown season.");
    }

    /**
     * Returns season with ID.
     *
     * @param id   ID
     * @param show show
     * @return season with ID
     */
    private static cz.vhromada.catalog.domain.Season getSeason(final Integer id, final cz.vhromada.catalog.domain.Show show) {
        for (final cz.vhromada.catalog.domain.Season season : show.getSeasons()) {
            if (id.equals(season.getId())) {
                return season;
            }
        }

        throw new IllegalStateException("Unknown season.");
    }

    /**
     * Updates season in show.
     *
     * @param show   show
     * @param season season
     */
    private static void updateSeason(final cz.vhromada.catalog.domain.Show show, final cz.vhromada.catalog.domain.Season season) {
        final List<cz.vhromada.catalog.domain.Season> seasons = new ArrayList<>();
        for (final cz.vhromada.catalog.domain.Season seasonDomain : show.getSeasons()) {
            if (seasonDomain.equals(season)) {
                seasons.add(season);
            } else {
                seasons.add(seasonDomain);
            }
        }
        show.setSeasons(seasons);
    }

}

File: src/main/kotlin/cz/vhromada/catalog/facade/impl/SeasonFacadeImpl.kt
package cz.vhromada.catalog.facade.impl

import cz.vhromada.catalog.entity.Season
import cz.vhromada.catalog.entity.Show
import cz.vhromada.catalog.facade.SeasonFacade
import cz.vhromada.common.facade.AbstractMovableChildFacade
import cz.vhromada.common.mapper.Mapper
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.utils.sorted
import cz.vhromada.common.validator.MovableValidator
import org.springframework.stereotype.Component

/**
 * A class represents implementation of facade for seasons.
 *
 * @author Vladimir Hromada
 */
@Component("seasonFacade")
class SeasonFacadeImpl(
        showService: MovableService<cz.vhromada.catalog.domain.Show>,
        accountProvider: AccountProvider,
        timeProvider: TimeProvider,
        mapper: Mapper<Season, cz.vhromada.catalog.domain.Season>,
        showValidator: MovableValidator<Show>,
        seasonValidator: MovableValidator<Season>
) : AbstractMovableChildFacade<Season, cz.vhromada.catalog.domain.Season, Show, cz.vhromada.catalog.domain.Show>(showService, accountProvider, timeProvider, mapper, showValidator,
        seasonValidator), SeasonFacade {

    override fun getDomainData(id: Int): cz.vhromada.catalog.domain.Season? {
        val shows = service.getAll()
        for (show in shows) {
            for (season in show.seasons) {
                if (id == season.id) {
                    return season
                }
            }
        }

        return null
    }

    override fun getDomainList(parent: Show): List<cz.vhromada.catalog.domain.Season> {
        return service.get(parent.id!!)!!.seasons
    }

    override fun getForAdd(parent: Show, data: cz.vhromada.catalog.domain.Season): cz.vhromada.catalog.domain.Show {
        val show = service.get(parent.id!!)!!
        val seasons = show.seasons.toMutableList()
        seasons.add(data)

        return show.copy(seasons = seasons)
    }

    override fun getForUpdate(data: Season): cz.vhromada.catalog.domain.Show {
        val show = getShow(data)
        val season = getDataForUpdate(data).copy(episodes = getSeason(data.id, show).episodes)

        return show.copy(seasons = updateSeason(show, season))
    }

    override fun getForRemove(data: Season): cz.vhromada.catalog.domain.Show {
        val show = getShow(data)
        val seasons = show.seasons.filter { it.id != data.id }

        return show.copy(seasons = seasons)
    }

    override fun getForDuplicate(data: Season): cz.vhromada.catalog.domain.Show {
        val show = getShow(data)
        val season = getSeason(data.id, show)
        val seasons = show.seasons.toMutableList()
        seasons.add(season.copy(id = null, episodes = season.episodes.map { it.copy(id = null) }))

        return show.copy(seasons = seasons)
    }

    @Suppress("DuplicatedCode")
    override fun getForMove(data: Season, up: Boolean): cz.vhromada.catalog.domain.Show {
        var show = getShow(data)
        val season = getSeason(data.id, show)
        val seasons = show.seasons.sorted()

        val index = seasons.indexOf(season)
        val other = seasons[if (up) index - 1 else index + 1]
        val position = season.position!!
        season.position = other.position
        other.position = position

        show = show.copy(seasons = updateSeason(show, season))
        return show.copy(seasons = updateSeason(show, other))
    }

    /**
     * Returns show for season.
     *
     * @param season season
     * @return show for season
     */
    private fun getShow(season: Season): cz.vhromada.catalog.domain.Show {
        for (show in service.getAll()) {
            for (seasonDomain in show.seasons) {
                if (season.id == seasonDomain.id) {
                    return show
                }
            }
        }

        throw IllegalStateException("Unknown season.")
    }

    /**
     * Returns season with ID.
     *
     * @param id   ID
     * @param show show
     * @return season with ID
     */
    private fun getSeason(id: Int?, show: cz.vhromada.catalog.domain.Show): cz.vhromada.catalog.domain.Season {
        for (season in show.seasons) {
            if (id == season.id) {
                return season
            }
        }

        throw IllegalStateException("Unknown season.")
    }

    /**
     * Updates season.
     *
     * @param show   show
     * @param season season
     * @return updated seasons
     */
    @Suppress("DuplicatedCode")
    private fun updateSeason(show: cz.vhromada.catalog.domain.Show, season: cz.vhromada.catalog.domain.Season): List<cz.vhromada.catalog.domain.Season> {
        val seasons = mutableListOf<cz.vhromada.catalog.domain.Season>()
        for (seasonDomain in show.seasons) {
            if (seasonDomain == season) {
                val audit = getAudit()
                season.audit = seasonDomain.audit!!.copy(updatedUser = audit.updatedUser, updatedTime = audit.updatedTime)
                seasons.add(season)
            } else {
                seasons.add(seasonDomain)
            }
        }

        return seasons
    }

}
